controller/deployer/strategies/main.go

```diff
@@ -7,20 +7,132 @@ import (
 	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
 	"github.com/flynn/flynn/controller/client"
 	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/pkg/attempt"
+	"github.com/flynn/flynn/pkg/cluster"
+	"github.com/flynn/flynn/pkg/stream"
 )
 
-type PerformFunc func(log15.Logger, *controller.Client, *ct.Deployment, chan<- ct.DeploymentEvent) error
+type UnknownStrategyError struct {
+	Name string
+}
+
+func (e UnknownStrategyError) Error() string {
+	return fmt.Sprintf(`deployer: unknown strategy "%s"`, e.Name)
+}
+
+type PerformFunc func(d *Deploy) error
 
 var performFuncs = map[string]PerformFunc{
 	"all-at-once": allAtOnce,
 	"one-by-one":  oneByOne,
 }
 
-func Get(strategy string) (PerformFunc, error) {
-	if f, ok := performFuncs[strategy]; ok {
-		return f, nil
+func Perform(d *ct.Deployment, client *controller.Client, deployEvents chan<- ct.DeploymentEvent, logger log15.Logger) error {
+	log := logger.New("fn", "Perform", "deplyment_id", d.ID, "app_id", d.AppID)
+
+	log.Info("validating deployment strategy")
+	performFunc, ok := performFuncs[d.Strategy]
+	if !ok {
+		err := UnknownStrategyError{d.Strategy}
+		log.Error("error validating deployment strategy", "err", err)
+		return err
+	}
+
+	deploy := &Deploy{
+		Deployment:      d,
+		client:          client,
+		deployEvents:    deployEvents,
+		logger:          logger,
+		newReleaseState: make(map[string]int, len(d.Processes)),
+		oldReleaseState: make(map[string]int, len(d.Processes)),
+		knownJobs:       make(map[string]struct{}),
+	}
+
+	log.Info("connecting job event stream")
+	if err := deploy.connectEventStream(); err != nil {
+		log.Error("error connecting job event stream", "err", err)
+		return err
+	}
+	defer deploy.closeEventStream()
+
+	log.Info("determining deployment state")
+	c, err := cluster.NewClient()
+	if err != nil {
+		log.Error("error determining deployment state", "err", err)
+		return err
 	}
-	return nil, fmt.Errorf("Unknown strategy '%s'!", strategy)
+	hosts, err := c.ListHosts()
+	if err != nil {
+		log.Error("error determining deployment state", "err", err)
+		return err
+	}
+	for _, host := range hosts {
+		for _, job := range host.Jobs {
```

Comment:
I avoided using the controller APIs because the test I wrote for this was all about deploying the controller, so if the deployer starts during a scheduler restart, there is potential for the deployer to get an incorrect view of actual jobs running (because the scheduler is still itself syncing with the cluster, so the fact that it itself is up may not yet be visible in the controller).

One way to get around that is to ensure that the scheduler leader is in sync with the cluster, then we can be more confident the controller has the correct view of what jobs are running.

WDYT?